pkg/controller/keyvault/gateway_kv_serviceaccount.go

```diff
@@ -0,0 +1,137 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+var kvSaControllerName = controllername.New("gateway", "keyvault", "serviceaccount")
+
+// KvServiceAccountReconciler reconciles the default "azure-app-routing-kv" ServiceAccount in each namespace where users
+// create a Gateway resource with the Keyvault cert TLS option. Users can tie this ServiceAccount to their own MSI via
+// workload identity through annotations on their Gateway resources.
+type KvServiceAccountReconciler struct {
+	client client.Client
+	events record.EventRecorder
+}
+
+func NewKvServiceAccountReconciler(mgr ctrl.Manager) error {
+	metrics.InitControllerMetrics(kvSaControllerName)
+
+	return kvSaControllerName.AddToController(ctrl.NewControllerManagedBy(mgr).
+		For(&gatewayv1.Gateway{}), mgr.GetLogger()).
+		Complete(&KvServiceAccountReconciler{
+			client: mgr.GetClient(),
+			events: mgr.GetEventRecorderFor("app-routing-operator"),
+		})
+
+}
+
+func (k *KvServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
+	var err error
+
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(kvSaControllerName, res, err)
+	}()
+
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return res, err
+	}
+	logger = kvSaControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	gwObj := &gatewayv1.Gateway{}
+	if err = k.client.Get(ctx, req.NamespacedName, gwObj); err != nil {
+		return res, client.IgnoreNotFound(err)
+	}
+
+	clientId, err := extractClientIdForManagedSa(gwObj)
+	if clientId == "" {
+		return res, err
+	}
+
+	if err != nil {
+		var userErr userError
+		if errors.As(err, &userErr) {
+			logger.Info("user error while extracting clientId from Gateway: %s", userErr.userMessage)
+			k.events.Event(gwObj, Warning.String(), "InvalidInput", userErr.userMessage)
+			return res, nil
+		}
+		logger.Error(err, "failed to extract clientId from Gateway object")
+		return res, err
+	}
+
+	toCreate := &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ServiceAccount",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      appRoutingSaName,
+			Namespace: req.Namespace,
+		},
+	}
+
+	logger.Info("checking for existing ServiceAccount")
+
+	existing := &corev1.ServiceAccount{}
+	err = k.client.Get(ctx, types.NamespacedName{Name: toCreate.Name, Namespace: toCreate.Namespace}, existing)
+	if client.IgnoreNotFound(err) != nil {
+		logger.Error(err, "failed to fetch existing app routing serviceaccount")
+		return res, err
+	}
+
+	if existing != nil && existing.Annotations != nil && existing.Annotations[wiSaClientIdAnnotation] != "" {
+		existingClientId := existing.Annotations[wiSaClientIdAnnotation]
+		if existingClientId != clientId {
+			errText := fmt.Sprintf("gateway specifies clientId %s but azure-app-routing-kv ServiceAccount already uses clientId %s", clientId, existingClientId)
```

Comment:
so it's impossible for them to change the client id for the service account? that doesn't seem correct. They technically could if they deleted the SA but that's a workaround

Reply:
They could modify the ServiceAccount first, right? We don't have any controllers watching the ServiceAccount itself

Reply:
Not sure I like this either, can lead to confusion if I edit a SA but not the gateway. Any reason why you're forcing the SA edit over allowing it to control the SA from the gateway regardless?

Reply:
I just think that since there's a many-to-one relatinonship between listeners and the app-routing service account that uses the clientId from that listener, it makes sense to have the source of truth be the "one" and not the "many"

Reply:
Or we could avoid this problem entirely by having user specify SA instead of specify client ID. I still think that's the better solution as it's more flexible and the behavior is straightforward

Reply:
how do we know which client ID customers want to use if they don't want to make their own serviceaccount? This was a specific ask from the PM side so I'm hesitant to just discard the managed SA option. I also just think that we should do whatever we can to make this experience as minimally intensive as possible given the number of steps required to set up WI

Reply:
We don't. The customer brings their own service account. It's not a worse experience, it's better because the behavior is well defined and it's more flexible.

Asking a customer to bring their own service account is not a large ask in any shape and makes the behavior more straightforward. Removing a single simple touch step isn't worth added complexity for a feature that should be straightforward

Reply:
I'll set up time with the PMs to talk this through -- this should've been brought up during the design and review process.

Reply:
What did we resolve to here? Customers bring SA for all gateway related tasks off the bat?

Reply:
yep exactly - resolving conversation accordingly